assets/js/atomic/blocks/product/image/block.js

```diff
@@ -2,101 +2,129 @@
-	) : null;
-};
-
-const Image = ( { layoutPrefix, loaded, image, onLoad } ) => {
-	const cssClass = classnames( `${ layoutPrefix }__product-image__image`, {
-		[ `${ layoutPrefix }__product-image__image_placeholder` ]:
-			! loaded && ! image,
-	} );
-	const { thumbnail, srcset, sizes, alt } = image || {};
-	return (
-		<Fragment>
-			{ image && (
-				<img
-					className={ cssClass }
-					src={ thumbnail }
-					srcSet={ srcset }
-					sizes={ sizes }
-					alt={ alt }
-					onLoad={ onLoad }
-					hidden={ ! loaded }
-				/>
-			) }
-			{ ! loaded && (
-				<img
-					className={ cssClass }
-					src={ PLACEHOLDER_IMG_SRC }
-					alt=""
-				/>
-			) }
-		</Fragment>
-	);
-};
+import ProductSaleBadge from '../sale-badge/block.js';
 
+/**
+ * Product Image Block Component.
+ *
+ * @param {Object} props                  Incoming props.
+ * @param {string} [props.className]      CSS Class name for the component.
+ * @param {boolean} [props.productLink]   Whether or not to display a link to the product page.
+ * @param {boolean} [props.showSaleBadge] Whether or not to display the on sale badge.
+ * @param {string} [props.saleBadgeAlign] How should the sale badge be aligned if displayed.
+ * @param {Object} [props.product]        Optional product object. Product from context will be used if
+ *                                        this is not provided.
+ * @return {*} The component.
+ */
 const ProductImage = ( {
 	className,
-	product,
 	productLink = true,
 	showSaleBadge = true,
 	saleBadgeAlign = 'right',
+	...props
 } ) => {
+	const productDataContext = { ...useProductDataContext(), ...props };
+	const { layoutStyleClassPrefix } = useInnerBlockLayoutContext();
+	const { product } = productDataContext;
+	const componentClass = `${ layoutStyleClassPrefix }__product-image`;
 	const [ imageLoaded, setImageLoaded ] = useState( false );
-	const { layoutStyleClassPrefix } = useInnerBlockConfigurationContext();
-	const image =
-		product.images && product.images.length ? product.images[ 0 ] : null;
 
-	const renderedSalesAndImage = (
-		<Fragment>
-			<SaleBadge
-				product={ product }
-				saleBadgeAlign={ saleBadgeAlign }
-				shouldRender={ showSaleBadge }
-			/>
-			<Image
-				layoutPrefix={ layoutStyleClassPrefix }
-				loaded={ imageLoaded }
-				image={ image }
-				onLoad={ () => setImageLoaded( true ) }
-			/>
-		</Fragment>
-	);
+	if ( ! product ) {
+		return (
+			<div
+				className={ classnames(
+					className,
+					componentClass,
+					'is-loading'
+				) }
+			>
+				<ImagePlaceholder componentClass={ componentClass } />
+			</div>
+		);
+	}
+
+	const image =
+		product?.images && product.images.length ? product.images[ 0 ] : null;
 
 	return (
-		<div
-			className={ classnames(
-				className,
-				`${ layoutStyleClassPrefix }__product-image`
-			) }
-		>
+		<div className={ classnames( className, componentClass ) }>
 			{ productLink ? (
 				<a href={ product.permalink } rel="nofollow">
-					{ renderedSalesAndImage }
```

Comment:
Just curious, why not keeping `renderedSalesAndImage`? It saved some duplicated lines of code.

Reply:
I found it a little hard to follow the logic before, where now it's in two clear blocks.